basic/bittwister_test.go

```diff
@@ -0,0 +1,609 @@
+
+	tt := []struct {
+		name             string
+		targetBandwidth  int64
+		tolerancePercent int
+	}{{
+		name:             "512 Kbps",
+		targetBandwidth:  512 * 1000,
+		tolerancePercent: 50,
+	},
+		{
+			name:             "1 Mbps",
+			targetBandwidth:  1024 * 1000,
+			tolerancePercent: 40,
+		},
+		{
+			name:             "2 Mbps",
+			targetBandwidth:  2 * 1024 * 1000,
+			tolerancePercent: 35,
+		},
+		{
+			name:             "4 Mbps",
+			targetBandwidth:  4 * 1024 * 1000,
+			tolerancePercent: 35,
+		},
+		{
+			name:             "8 Mbps",
+			targetBandwidth:  8 * 1024 * 1000,
+			tolerancePercent: 35,
+		},
+		{
+			name:             "16 Mbps",
+			targetBandwidth:  16 * 1024 * 1000,
+			tolerancePercent: 35,
+		},
+		{
+			name:             "32 Mbps",
+			targetBandwidth:  32 * 1024 * 1000,
+			tolerancePercent: 35,
+		},
+	}
+
+	for _, tc := range tt {
+		tc := tc
+		t.Run(tc.name, func(t *testing.T) {
+			err = iperfServer.SetBandwidthLimit(tc.targetBandwidth)
+			require.NoError(t, err, "Error setting bandwidth limit")
+
+			t.Log("Starting bandwidth test. It takes a while.")
+			startTime := time.Now()
+			ctx, cancel := context.WithTimeout(context.Background(), commandTimeout)
+			defer cancel()
+			output, err := iperfClient.ExecuteCommandWithContext(ctx,
+				"iperf3", "-c", iperfServerIP,
+				"-t", fmt.Sprint(int64(iperfTestDuration.Seconds())),
+				"-P", fmt.Sprint(iperfParallelClients), "--json")
+			require.NoError(t, err, "Error executing command")
+
+			elapsed := time.Since(startTime)
+			t.Logf("Test took %d seconds", int64(elapsed.Seconds()))
+
+			var iperfOutput struct {
+				End struct {
+					SumReceived struct {
+						BitsPerSecond float64 `json:"bits_per_second"`
+					} `json:"sum_received"`
+				} `json:"end"`
+			}
+			err = json.Unmarshal([]byte(output), &iperfOutput)
+			require.NoError(t, err, "Error unmarshalling JSON")
+
+			deviationPercent := math.Abs(iperfOutput.End.SumReceived.BitsPerSecond-float64(tc.targetBandwidth)) / float64(tc.targetBandwidth) * 100
+			assert.LessOrEqual(t, deviationPercent, float64(tc.tolerancePercent), "Deviation is too high")
+
+			t.Logf("Bandwidth expected: %v \tgot: %v \tdeviation: %v%% \ttolerance: %v%%",
+				formatBandwidth(float64(tc.targetBandwidth)),
+				formatBandwidth(iperfOutput.End.SumReceived.BitsPerSecond),
+				math.Round(deviationPercent),
+				tc.tolerancePercent)
+		})
+	}
+}
+
+func TestBittwister_Packetloss(t *testing.T) {
+	t.Parallel()
+	// Setup
+
+	const (
+		numOfPingPackets = 1000
+		packetTimeout    = 10 * time.Millisecond
+		commandTimeout   = 10 * time.Minute
+	)
+
+	mother, err := knuu.NewInstance("mother")
+	require.NoError(t, err, "Error creating instance")
+
+	require.NoError(t, mother.EnableBitTwister(), "Error enabling BitTwister")
+	err = mother.SetImage(gopingImage)
+	require.NoError(t, err, "Error setting image")
+
+	gopingPort := 8001
+	err = mother.SetEnvironmentVariable("SERVE_ADDR", fmt.Sprintf("0.0.0.0:%d", gopingPort))
```

Comment:
optimization: If you set the env variable after you commit, the Instance is prepared faster, as it does not need to build a container image.

Reply:
done

Reply:
done